### ISSUE TITLE
Mixing TabRouter and StackRouter

### ISSUE THREAD

**Issue by wer-mathurin** [Milad-Akarie/auto_route_library#430]
Just a quick question.

Is it possible to mix TabRouter with StackRouter.

The use case I'm looking for:
Base on  the example provided in the repository.

Navigate to page with the tabrouter....so far so good.
But If I want to navigate to a detail page in one of the selected page,  I would like the AppBar to reflect that I pushed a new route updated with the back arrow.  If I'm nesting a scaffold I can see the route is pushed...but I will make more sense to change the AppBar in the HomePage.

In the example bellow, from the home page, I have tile that will open detail pages.

![image](https://user-images.githubusercontent.com/39314023/112700782-4aa37700-8e65-11eb-87ee-3c927c4fca39.png)

As you can see, navigating to the detail page, by nesting a scaffold we can see the route is pushed.  But the objective is to show the nested route using the original AppBar.  Is it possible, I can't figure it out using the TabRouter

![image](https://user-images.githubusercontent.com/39314023/112700898-8e967c00-8e65-11eb-9b02-c03dcec30047.png)

Probably If I'm using the StackRouter and manage the routing using replace() when clicking on the navigationbaritem, And use a scaffold in each page, I will be able to achieve my goal, but just wondering if I can achieve this using the very neat AutoTabRouter.

LMM

**Comment by Milad-Akarie** [comment]
Hello @wer-mathurin  I believe the best you can do in your case is have each tab implement it's own AppBar?

**Comment by wer-mathurin** [comment]
Thanks for your response.

Let me clarify the use case.
I would like to see if this is possible to use the tabrouter for having a
persistent BottomAppBar that will *replace* the current route and having a
AppBar that will reflect the state of the navigation like the stack router
do, so when I push a child page will show the back icon.


I think the only possible way with the current implementation of this
library is to use the stack router.and implement the BottomAppBar
navigation myself.  I'm right?

Lmm



Le dim. 28 mars 2021 05 h 33, Milad akarie ***@***.***> a
écrit :

**Comment by Milad-Akarie** [comment]
@wer-mathurin Could you please share your router setup so I can get a better  idea of what you're trying to achieve? 
TabRouters don't replace their children, all children are pushed in parallel, it's internally using an indexedStack.

or did you mean pushing a child route to one of your tabs and have the Main Appbar reflect the state?

**Comment by wer-mathurin** [comment]
@Milad-Akarie

did you mean pushing a child route to one of your tabs and have the Main
Appbar reflect the state?
*YES !*




@CustomAutoRouter(
  replaceInRouteName: 'Page,Route',
  // transitionsBuilder: TransitionsBuilders.fadeIn,
  routes: <AutoRoute>[
    searchRoute,
    AutoRoute(
      path: '/',
      page: HomeRouter,
      usesTabsRouter: true,
      children: [
        RedirectRoute(path: '', redirectTo: 'home'),
        AutoRoute(
          path: 'home',
          page: EmptyRouterPage,
          name: 'homeTab',
          children: [
            AutoRoute(path: '', page: HomePage),
            AutoRoute(path: 'information', page: InformationCategoryPage),
          ],
        ),
      ],
    ),
  ],
)
class $AppRouter {}

**Comment by Milad-Akarie** [comment]
@wer-mathurin I'm actually working on something to handle navigation change cases like yours, it's not complete yet but I guess you could give it a try
copy this widget to your source code

```dart
typedef NavigationChangeBuilder = Widget Function(BuildContext context, RoutingController topMostRouter);

class NavigationListener extends StatefulWidget {
  final NavigationChangeBuilder builder;

  const NavigationListener({Key? key, required this.builder}) : super(key: key);

  @override
  _NavigationListenerState createState() => _NavigationListenerState();
}

class _NavigationListenerState extends State<NavigationListener> {
  bool _didInit = false;

  @override
  Widget build(BuildContext context) {
    return widget.builder(
      context,
      RootRouterDelegate.of(context).controller.topMost,
    );
  }

  void _listener() {
    setState(() {});
  }

  @override
  void deactivate() {
    super.deactivate();
    RootRouterDelegate.of(context).removeListener(_listener);
  }

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    if (!_didInit) {
      _didInit = true;
      RootRouterDelegate.of(context).addListener(_listener);
    }
  }
}
```
wrap your appbar with it 

```dart
        Scaffold(
            appBar: PreferredSize(
              preferredSize: const Size.fromHeight(kToolbarHeight),
              child: NavigationListener(builder: (context, router) {
              // router is the topmost router in stack ( the visible router )
                return AppBar(
                  title: Text(router.current?.name ?? ''),
                  leading: (router is StackRouter && router.stack.length > 1)
                      ? IconButton(
                          icon: BackButtonIcon(),
                          onPressed: router.pop,
                        )
                      : null,
                );
              }),
            ),
           ....
          );
```

**Comment by wer-mathurin** [comment]
Thank you, that's a very good solution.
This is working for me.

I will add a check to the condition:
(router is StackRouter && router.stack.length > 1 *&& router != router.root
**)*

In the example I provided, this will prevent showing the leading indicator
when navigating the searchPage that is showing on top of the tab
router...otherwize you can see the leading getting rendered while the
searchPage is sliding from left to right in my use case.

**Comment by Milad-Akarie** [comment]
@wer-mathurin  That's good news. 
let me know if you have any suggestions on how we can enhance this.

**Comment by herman-the-worm** [comment]
@Milad-Akarie was this feature implemented in v2.0 already? I cannot find a reference to it in the documentation GitHub website
